### PR TITLE
Allows navigation to parklets through a sharable link

### PR DESCRIPTION
**Pull request by mikehern**
Parklets can be accessed by navigating to `/parkletName`.  The parklet path is derived by taking the raw parklet data, and stripping out punctuation from its title.  Note: this uses [catch-all routing](https://stackoverflow.com/a/36623117/9420872) on the server, so that if the user goes to an invalid url, they're presented the map page with a message saying the link is unrecognized.

When the user goes to a parklet link, they are taken to a modal containing detail data.

- Temporarily displays a generated link when clicking on a marker
- Temporarily includes a close button for the modal until the feature is complete
- Temporarily renders a subset of parklet detail data
- Disables test to catch an unrecognized url on the server, and defers to client for testing
- Contains a helper function to generate links for each parklet
- Handles going to parklet links without first requiring navigation to the home page
- Does NOT handle browser address updates, nor updates the browser navigation history

[Trello Card](https://trello.com/c/g79JYASw)

### CODE REVIEW

File: server/server.js [hunk 1]
```diff
@@ -13,4 +13,12 @@ useMorganMiddleware(app);
 app.use(express.static(assets));
 app.use('/test', router);
 
+app.get('/*', (req, res) => {
+  res.sendFile(path.join(__dirname, '../client/'), (err) => {
```
Comment (jcee-io) [inline server/server.js:17]: I believe __dirname gives you the root of your whole file system if I'm not mistaken, which brings up the question of using '../client/' syntactically. No qualms if intended behavior is produced.

Comment (andino93) [reply server/server.js:17]: You still need to point to the next part of the path from the relative position of the file its being called in.

Comment (andino93) [reply server/server.js:17]: The alternative is to use `path.resolve()`

**Review by jcee-io — DISMISSED** [review]
Looks solid. Approved.

File: server/server.js (hunk 1 above)
Comment (mikehern) [reply server/server.js:17]: __dirname does give the root, and is the first argument of join.  `../client/` is supplied as the second argument.

My understanding is `join` concatenates all of the arguments, in order. The outcome is the server responds with the content in that concatenated filepath.

Comment (andino93) [reply server/server.js:17]: That is the correct understanding. If there is some uncertainty, print the asset path with and without the `../` and see how it changes the path

File: client/src/linkValidator.jsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import LinkedModal from './linkedModal';
+import makeUrlFriendly from '../utils/makeUrlFriendly';
+
+class LinkValidator extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { parklets: '', ...props };
+  }
+
+  componentWillReceiveProps(nextProps) {
```
Comment (andino93) [inline client/src/linkValidator.jsx]: Can you update to use new lifecycle hook? `componentWIllReceiveProps` will be deprecated soon

https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops

File: server/server.js [hunk 3]
```diff
@@ -13,4 +13,12 @@ useMorganMiddleware(app);
 app.use(express.static(assets));
```
Comment (andino93) [inline server/server.js:13]: Do we need both the `/*` fallback for client side routing and express.static?

File: server/server.js (hunk 1 above)
Comment (andino93) [inline server/server.js:17]: use `assets` declaration or remove it. Repeated code otherwise.

**Review by andino93 — CHANGES_REQUESTED** [review]

File: spec/server/server.spec.js [hunk 4]
```diff
@@ -2,15 +2,10 @@ import request from 'supertest';
 
 import server from '../../server/server';
 
-describe('App Server', () => {
+describe('App Server', () => {
```
Comment (andino93) [inline spec/server/server.spec.js:5]: This test is necessary with how the server catches all things to client side render. Perhaps a test to ensure we're getting back static assets instead would be more meaningful

**Review by andino93 — CHANGES_REQUESTED** [review]

File: client/src/linkValidator.jsx (hunk 2 above)
Comment (mikehern) [reply client/src/linkValidator.jsx]: This is for React 16.3.1 and we're on 16.2

Are we all OK with the upgrade? If so, I can make the update and update to the new lifecycle method.

File: server/server.js (hunk 1 above)
Comment (mikehern) [reply server/server.js:17]: Addresses the other comment too^

I can't get react-router to handle the catch-all fallback unless both exist.

I've tried the following:

Removing the express.static code, leaving the catch-all block, and it won't render the map at all.

Replacing the argument within `sendFile` with `assets`.  The browser shows the following output: `{"code":"EISDIR"}`.

Not sure exactly what to do here, but if you have a working solution, I could use some help.

File: spec/server/server.spec.js (hunk 4 above)
Comment (mikehern) [reply spec/server/server.spec.js:5]: Cool - sounds like a reasonable test.  Thanks for the suggestion, will make the update.

**Pull request closed without merging**